File: leetcode/solutions/0454.4SumII/fourSumCount.go
package solution

func fourSumCount(A []int, B []int, C []int, D []int) int {
	sumCountMap := make(map[int]int)

	for _, c := range C {
		for _, d := range D {
			sumCountMap[c+d]++
		}
	}

	var count int

	for _, a := range A {
		for _, b := range B {
			count += sumCountMap[-a-b]
		}
	}

	return count
}

File: leetcode/solutions/0454.4SumII/fourSumCount.ts
export default function fourSumCount(
  A: number[],
  B: number[],
  C: number[],
  D: number[]
): number {
  const map = new Map<number, number>();

  for (const a of A) {
    for (const b of B) {
      const sum = a + b;
      map.set(sum, (map.get(sum) || 0) + 1);
    }
  }

  let count = 0;

  for (const c of C) {
    for (const d of D) {
      const sum = -c - d;
      count += map.get(sum) || 0;
    }
  }

  return count;
}
